nano/node/node.cpp

```diff
@@ -510,19 +605,52 @@ class network_message_visitor : public nano::message_visitor
 	{
 		if (node.config.logging.network_message_logging ())
 		{
-			BOOST_LOG (node.log) << boost::str (boost::format ("Confirm_req message from %1% for %2%") % sender % message_a.block->hash ().to_string ());
+			if (!message_a.roots_hashes.empty ())
+			{
+				BOOST_LOG (node.log) << boost::str (boost::format ("Confirm_req message from %1% for hashes:roots %2%") % sender % message_a.roots_string ());
+			}
+			else
+			{
+				BOOST_LOG (node.log) << boost::str (boost::format ("Confirm_req message from %1% for %2%") % sender % message_a.block->hash ().to_string ());
+			}
 		}
 		node.stats.inc (nano::stat::type::message, nano::stat::detail::confirm_req, nano::stat::dir::in);
 		node.peers.contacted (sender, message_a.header.version_using);
 		// Don't load nodes with disabled voting
 		if (node.config.enable_voting)
 		{
 			auto transaction (node.store.tx_begin_read ());
-			auto successor (node.ledger.successor (transaction, nano::uint512_union (message_a.block->previous (), message_a.block->root ())));
-			if (successor != nullptr)
+			if (message_a.block != nullptr)
 			{
-				auto same_block (successor->hash () == message_a.block->hash ());
-				confirm_block (transaction, node, sender, std::move (successor), !same_block);
+				auto successor (node.ledger.successor (transaction, nano::uint512_union (message_a.block->previous (), message_a.block->root ())));
+				if (successor != nullptr)
+				{
+					auto same_block (successor->hash () == message_a.block->hash ());
+					confirm_block (transaction, node, sender, std::move (successor), !same_block);
+				}
+			}
+			else if (!message_a.roots_hashes.empty ())
+			{
+				std::vector<nano::block_hash> blocks_bundle;
+				for (auto root_hash : message_a.roots_hashes)
```

Comment:
Maybe `auto &` to save an array copy, possibly same in other auto range-based loops.